### ISSUE TITLE
Temporary file issue with r-lintr

### ISSUE THREAD

**Issue by ashiklom** [flycheck/flycheck#1634]
## Bug description
When running in the background, the `r-lintr` checker is unable to reconcile the correct R source path from the temporary one, so Flycheck error messages appear like this:

```
In "../../../private/var/folders/4r/94db7f8j12q1xyjzyrtj6hbwl14jnc/T/flycheck09Z4SY/flycheck.R":
unexpected end of input
```

...and highlight to the first line of the buffer, regardless of where in the buffer the error actually occurs.

## Steps to reproduce
1. With a clean Emacs instance, install and `load-library` the latest `ess` and `flycheck` from Melpa.
2. Open an R script with some style and code errors in it.
3. Disable flymake: `M-x flymake-mode`
4. Enable flycheck: `M-x flycheck-mode`
5. Wait for flycheck to run.

An error like the one above should appear on the first line, _regardless_ of where the _actual_ error is in the file.

Note that `flycheck-compile` works fine -- it spits out the right errors _and_ style mistakes in all the right places.

## Screenshots

<img width="600" alt="Screen Shot 2019-10-18 at 4 55 44 PM" src="https://user-images.githubusercontent.com/5680221/67127856-2c454080-f1c8-11e9-83d2-d473d3568190.png">

## System configuration
```
Syntax checkers for buffer flycheck.R in ess-r-mode:

First checker to run:

  r-lintr
    - may enable:    yes
    - predicate:     t
    - executable:    Found at /usr/local/bin/R
    - lintr library: present

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20190930.1425)
Emacs version:    26.2
System:           x86_64-apple-darwin14.5.0
Window system:    ns
```

Vanilla Emacs, with minimal configuration, on MacOS High Sierra. `.emacs.d/init.el` is:

```emacs-lisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))
(add-to-list 'package-archives '("melpa-stable" . "https://stable.melpa.org/packages/"))
(package-initialize)

(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(package-selected-packages (quote (julia-mode ess flycheck))))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )
```

Emacs configuration:
- [X] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

**Comment by tviti** [comment]
Just chiming in to say that I'm also experiencing this issue (also using plain Emacs on macOS). I'm actually not 100% convinced that the issue is rooted in flycheck though. I have been bouncing between flymake and flycheck recently, and noticed that they are both misbehaving w/ lintr atm. Also have noticed that flymake seems to working just fine on my office PC (running centos 7).

**Comment by tviti** [comment]
Some debugging: I rolled back `lintr` as follows:

```
> require(devtools)
> install_version("lintr", version = "1.0.3")
```

and it seems to sort this out (linting in `ess-r-mode` and `Poly-markdown+R` are
behaving normally again), so I guess it's somewhere in the `lintr` update from
`v1.0.3` to `v2.0 `that things break (which I guess makes sense, since it is a
major version update after all). I'm guessing that this isn't an issue on my
centos box simply because it hasn't gotten the upgrade.

**Comment by ashiklom** [comment]
Reverting `lintr` from 2.0 to 1.0.3 fixes the problem. I'll have to look carefully at differences in the output between the two versions to see what might be causing the problem.

**Comment by cpitclaudel** [comment]
Thanks a lot

**Comment by cpitclaudel** [comment]
Any news on this @ashiklom?

**Comment by ashiklom** [comment]
OK, I think I found the issue:

The old version of `lintr` (1.0.3) only printed the relative path of the file in the error message:

```
scratch/lintr.R:1:12: style: Variable and function names should be all lowercase.
Problem <- ThisFunction()
           ^~~~~~~~~~~~
scratch/lintr.R:3:6: error: unexpected symbol
This is an error
     ^
```

The new version prints the absolute path to the file:

```
/Users/ashiklom/scratch/lintr.R:1:1: style: Variable and function name style should be snake_case.
Problem <- ThisFunction()
^~~~~~~
/Users/ashiklom/scratch/lintr.R:3:6: error: unexpected symbol
This is an error
     ^
```

I don't know enough about the internals of Flycheck to propose an exact bugfix. But my guess is that, at least for the `r-lintr`, it should involve stripping the temporary directory path from the relevant part of the error message and possibly replacing it with the path to the original file (not the temporary copy).

**Comment by cpitclaudel** [comment]
> But my guess is that, at least for the r-lintr, it should involve stripping the temporary directory path from the relevant part of the error message and possibly replacing it with the path to the original file (not the temporary copy).

That surprising, since Flycheck should already do that…

Do you have a concrete file that I could reproduce the issue with?  I installed a fresh copy of ESS and Flycheck in a clean Emacs, and errors are working fine.  Running `packageVersion("lintr")` says `‘2.0.1’`

**Comment by ashiklom** [comment]
Very strange! I'm running vanilla Emacs 26.3 on MacOS Catalina 10.15.3. ESS version 20200314.113. Flycheck version 20200318.1319.

My file is located in my home directory (`~/lintr.R`) and looks like this:

```r
# A comment

invalid syntax here
```

Flymake correctly highlights the syntax error on the word `syntax` ("unexpected symbol").

<img width="707" alt="Screen Shot 2020-03-19 at 11 55 46 AM" src="https://user-images.githubusercontent.com/5680221/77086758-93199080-69d8-11ea-9f5b-13dc57f333ce.png">

However, disabling flymake and enabling flycheck moves the error to the first line and gives the error described in this issue.

<img width="707" alt="Screen Shot 2020-03-19 at 11 54 34 AM" src="https://user-images.githubusercontent.com/5680221/77086635-6b2a2d00-69d8-11ea-8b6b-9ddb6e6244ec.png">

**Comment by cpitclaudel** [comment]
Most vexing: this is what I see:

![Screenshot from 2020-03-19 12-19-10](https://user-images.githubusercontent.com/8181630/77089285-8b201780-69fd-11ea-8b51-300fd37d8db2.png)

**Comment by ashiklom** [comment]
I wonder if this has to do with differences in how MacOS vs. Linux do temporary files and directories. I'll try to investigate when I have the time.

**Comment by cpitclaudel** [comment]
Yup, I expect that's it.  Can you try to run the following with `M-:` after enabling `flycheck-mode`, then run `M-x flycheck-buffer`?

```
(progn
  (defun ~/log-inputs (&rest args)
    (message "Called with %S" args))
  (advice-add 'flycheck-fix-error-filename :before #'~/log-inputs))
```

What does this print in `*Messages*`?

**Comment by ashiklom** [comment]
Thanks for the quick reply! Here is the full "Messages" buffer output (in case some of the earlier stuff is relevant too):

```
For information about GNU Emacs and the GNU system, type C-h C-a.
lintr.R has auto save data; consider M-x recover-this-file
Flymake mode disabled in current buffer
Flycheck mode enabled in current buffer
In "../../private/var/folders/t_/fn5c3s75553cv_zs9dzymt5c0000gp/T/flycheckPymCyl/lintr.R":
unexpected symbol
Quit
Making completion list...
nil
You can run the command ‘flycheck-buffer’ with C-c ! c
Called with (#s(flycheck-error #<buffer lintr.R> r-lintr "/private/var/folders/t_/fn5c3s75553cv_zs9dzymt5c0000gp/T/flycheckMSu2cK/lintr.R" 3 9 "unexpected symbol" error nil nil nil nil) ("/var/folders/t_/fn5c3s75553cv_zs9dzymt5c0000gp/T/flycheckMSu2cK/lintr.R" "/var/folders/t_/fn5c3s75553cv_zs9dzymt5c0000gp/T/flycheckMSu2cK") "/Users/ashiklom/")
```

**Comment by ashiklom** [comment]
Ah, OK, I just noticed something else that may contribute to the problem: The filename returned by `lintr` starts with `/private/var`, whereas the actual path just starts with `/var/` -- no `/private`. From the command line:

```
> pwd
/var/folders/t_/fn5c3s75553cv_zs9dzymt5c0000gp/T/flycheckbB2POo
> ls
lintr.R
> Rscript -e "lintr::lint('lintr.R')"
/private/var/folders/t_/fn5c3s75553cv_zs9dzymt5c0000gp/T/flycheckbB2POo/lintr.R:3:9: # error: unexpected symbol
invalid syntax here
        ^
```

After a bit more digging, it looks like the real path is actually `/private/var`, but MacOS automatically creates a symlink to `/var`.

```
> ls -l / | grep var
lrwxr-xr-x@  1 root  admin    11 Nov  8 18:42 var -> private/var
``` 

The Emacs equivalent to R's `normalizePath` is `file-truename`, so I can make `temporary-file-directory` return what I want with `(advice-add 'temporary-file-directory :filter-return #'file-truename)`, but that doesn't seem to fix Flycheck: The error message starts with `In "../../private/var/folders...`, so it registered the `/private`. 🤷‍♂

**Comment by cpitclaudel** [comment]
Thanks, that's good progress.  Let's make one more experiment then.  Can you add this snippet now?

```
(progn
  (defun ~/log-inputs (&rest args)
    (message "Called with %S" args))
  (defun ~/log-output (output)
    (message "Returned %S" output)
    output)
  (advice-add 'flycheck-same-files-p :before #'~/log-inputs)
  (advice-add 'flycheck-same-files-p :filter-return #'~/log-output))
```

This will tell us whether Flycheck sees that these are the same files.

**Comment by ashiklom** [comment]
Whoops, just saw this, sorry! I implemented the fix in #1698 on my branch and it resolved the issue. Not sure if there are any unintended side-effects, though, but I don't think there should be, based on my understanding of what `flycheck-same-files-p` is meant to do.

**Comment by cpitclaudel** [comment]
> Not sure if there are any unintended side-effects

Performance is the main one.  This is a very costly test in a hot function.

**Comment by ashiklom** [comment]
An inelegant (and inefficient? My Lisp is not good -- I'm very open to suggestions!) workaround for MacOS users is to add code like the following to your local configuration:

```emacs-lisp
(if (eq system-type 'darwin)
  ;; Doom users can replace `with-eval-after-load` with `after!`
  ;; and unquote `flycheck`
    (with-eval-after-load "flycheck"
      (defun flycheck-fix-macos-file (file-x)
        (if (string-match-p "^/var" file-x)
            (string-join (list "/private" file-x))
          file-x))
      (defun flycheck-fix-macos-files (file-xs)
        (mapcar #'flycheck-fix-macos-file file-xs))
      (advice-add 'flycheck-same-files-p :filter-args #'flycheck-fix-macos-files)))
```

This will automatically prepend `/private` to file paths starting with `/var`, which will allow flycheck to correctly recognize them as the same file.
